Update telemetry alongside latest specifications

whitfin opened the pull request:
This PR will update the telemetry formats to the latest specifications, and point to the latest endpoints. The old endpoints and formats are still accepted, but we want to move everything to the new structures to reduce complexity.

It also introduces crash reporting by sending error messages and stack traces in a `crash.report` event (which uses the internal Crash Analytics system). I'm not sure if how I wired this up is the best way for the daemon though, so open to suggestions!

cb1kenobi commented on packages/appcd-core/src/server.js:
```diff
@@ -162,6 +162,15 @@ export default class Server {
+			await Dispatcher.call('/appcd/telemetry', {
+				event: 'crash.report',
+				message: err.message,
+				stack: [ err.stack ]
```
`err.stack` is a string. Why wrap it in an array instead of just doing `err.stack.split(/\r\n|\n/)`?

whitfin replied:
This is the format of crash events used in other projects (as you can supply multiple stack traces).

cb1kenobi commented on packages/appcd-core/src/server.js:
```diff
@@ -162,6 +162,15 @@ export default class Server {
 
+		// report crash telemetry
+		process.on('uncaughtException', async err => {
+			await Dispatcher.call('/appcd/telemetry', {
```
No need to make this function `async` as we just want to fire and forget.

cb1kenobi commented on packages/appcd-telemetry/src/telemetry.js:
```diff
@@ -42,9 +42,9 @@ export default class Telemetry extends Dispatcher {
-		const aguid = cfg.get('telemetry.guid');
-		if (!aguid || typeof aguid !== 'string') {
-			throw new Error('Config is missing a required, valid "telemetry.guid"');
+		const app = cfg.get('telemetry.guid') || cfg.get('telemetry.app');
```
We should favor `telemetry.app` over `telemetry.guid`. I don't see how this would affect backwards compatibility.

whitfin replied:
If someone has already set `telemetry.guid`, it'll start being ignored in the future if that's not first. It might be that we don't care about that, but it's technically breaking.

cb1kenobi replied:
I don't believe anybody is using appcd-telemetry other than the daemon. `guid` will only be ignored if `app` is set, which is what consumers should be using anyways. The problem is if someone has `app` and `guid` set at the same time, which shouldn't happen, in which case we want them to use `app`, right?

cb1kenobi commented on packages/appcd-telemetry/src/telemetry.js:
```diff
@@ -127,9 +121,9 @@ export default class Telemetry extends Dispatcher {
-				platform: process.platform
+				version,
+				name: 		name || process.platform,
+				arch: 		process.arch
```
Don't use `process.arch`. Instead, import `arch` from `appcd-util` (we already import `osInfo`) and call `arch()` to determine the correct machine architecture (not the architecture of the compiled Node binary).

cb1kenobi commented on packages/appcd-telemetry/src/telemetry.js:
```diff
@@ -166,19 +160,25 @@ export default class Telemetry extends Dispatcher {
+				distribution: {
+					environment: 	this.environment,
+					version: 		this.version
+				},
```
No hanging comma.

cb1kenobi commented on packages/appcd-telemetry/src/telemetry.js:
```diff
@@ -166,19 +160,25 @@ export default class Telemetry extends Dispatcher {
+					id: 			this.sessionId
+				},
+				distribution: {
+					environment: 	this.environment,
```
Weird indentation.

whitfin commented:
@cb1kenobi I've removed the uncaughtException handler for now. I'll look at adding the crash endpoint when I get a sec. I've addressed all other comments, so should be good if you're cool with the changes.
